extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala

```diff
@@ -35,7 +35,7 @@ import Scalaz._
 import com.typesafe.scalalogging.StrictLogging
 import scalaz.syntax.tag._
 
-class Extractor[T <: Target](config: ExtractorConfig, target: T) extends StrictLogging {
+class Extractor[T <: Target](config: ExtractorConfig, target:T, writerSupplier: (ExtractorConfig, Target, String) => Writer) extends StrictLogging {
```

Comment:
This makes it so you don't have to unseal `Target`, which is unsound anyway, and your custom `writerSupplier` can simply take your specific target-that-is-not-a-`Target`.

```suggestion
class Extractor[T](config: ExtractorConfig, target: T)(writerSupplier: (ExtractorConfig, T, String) => Writer = Writer(_, _, _)) extends StrictLogging {
```

Reply:
Incidentally, @leo-da do you think this `String` should be `domain.LedgerId` instead, to make this signature more precise?

Reply:
yes, I would prefer `com.digitalasset.ledger.api.domain.LedgerId`. It would definitely make this type more precise:
```
(ExtractorConfig, Target, String) => Writer
```
However, PostgreSQLWriter is defined with `String` for ledgerId:
```
class PostgreSQLWriter(config: ExtractorConfig, target: PostgreSQLTarget, ledgerId: String)
```
I would say we can/should refactor. Does not have to be part of this PR.

Reply:
Thanks a lot for the review comments.
About the new signature, the way default Writer is provided does not compile and I tried a few more combinations but could not find a non-verbose way to provide a default Writer. For now, I have excluded and it and all clients are providing the Writer explicitly. 

For LedgerId, I will file a follow-up PR for it.

Reply:
Few other questions, not sure if this is the correct place :

1. Is the multi-table format still supported? If not, is it ok to remove the code in a new PR?
2. Is there a way to write to a different SQL database without completely re-writing the Writer implementation for it. Can some parts of PostgreSQL implementation be re-used?
Current PostgreSQLWriter uses Queries object for all the queries. 
If we rename PostgreSQLWriter to SQLWriter and pass different implementation of Queries based on dialect, then PostgreSQLWriter code can be reused?

Reply:
> About the new signature, the way default Writer is provided does not compile and I tried a few more combinations but could not find a non-verbose way to provide a default Writer.

I tested before suggesting and it worked; can you share the compile error?

Reply:
I get this error on compiling
```
extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala:37: error: missing parameter type for expanded function ((x$1: com.digitalasset.extractor.config.ExtractorConfig, x$2: <error>, x$3: String) => Writer(x$1, x$2, x$3))
class Extractor[T](config: ExtractorConfig, target: T)(writerSupplier: (ExtractorConfig, T, String) => Writer = Writer(_, _, _)) extends StrictLogging {

```

Reply:
Hi @S11001001 Please let me know if I am doing something wrong here.

Reply:
My bad, I forgot another important change to make this style work, which the error message pointed out to me; thanks. I pushed a commit to your branch to make this change.

Reply:
Thanks @S11001001 for merging the PR. Can you please answer the questions here, or point me to the right channel?
https://github.com/digital-asset/daml/pull/2366#discussion_r310417886

Reply:
> Few other questions, not sure if this is the correct place :

@amanmanocha Best place for questions is probably #daml-contributors (for questions about this repo's implementation or contributing) or #public (otherwise) on damldriven.slack.com.

> 1. Is the multi-table format still supported?

Yes.

> 2. Is there a way to write to a different SQL database without completely re-writing the Writer implementation for it. Can some parts of PostgreSQL implementation be re-used? Current PostgreSQLWriter uses Queries object for all the queries. If we rename PostgreSQLWriter to SQLWriter and pass different implementation of Queries based on dialect, then PostgreSQLWriter code can be reused?

My guess is that the abstract interface of `Queries` would be _ad hoc_, and thus unmaintainable locally. If things can be parameterized in a simpler way, then that would probably be fine. I can't say more specifically while keeping this response simple, though.